parachain/primitives/router/src/inbound/mod.rs

```diff
@@ -70,135 +69,158 @@ pub enum Destination {
+}
 
-		let create_instructions = |origin_location: Junction| -> Vec<Instruction<()>> {
-			vec![
-				UniversalOrigin(GlobalConsensus(network)),
-				DescendOrigin(X1(origin_location)),
-				WithdrawAsset(buy_execution_fee.clone().into()),
-				BuyExecution { fees: buy_execution_fee.clone(), weight_limit: Unlimited },
-				SetAppendix(
-					vec![
-						RefundSurplus,
-						DepositAsset {
-							assets: buy_execution_fee.into(),
-							beneficiary: (
-								Parent,
-								Parent,
-								GlobalConsensus(network),
-								origin_location,
-							).into(),
-						},
-					]
-					.into(),
-				),
-			]
-		};
+pub trait ConvertMessage {
+	fn convert(message: VersionedMessage) -> Result<Xcm<()>, ConvertMessageError>;
+}
 
-		match self {
-			Command::RegisterToken { gateway, token, create_call_index } => {
-				let owner = GlobalConsensusEthereumAccountConvertsFor::<[u8; 32]>::from_params(
-					&chain_id,
-					gateway.as_fixed_bytes(),
-				);
-
-				let origin_location = Junction::AccountKey20 { network: None, key: gateway.into() };
-
-				let asset_id = Self::convert_token_address(network, gateway, token);
-
-				let mut instructions = create_instructions(origin_location);
-				instructions.extend(vec![
-					Transact {
-						origin_kind: OriginKind::Xcm,
-						require_weight_at_most: Weight::from_parts(400_000_000, 8_000),
-						call: (
-							create_call_index,
-							asset_id,
-							MultiAddress::<[u8; 32], ()>::Id(owner),
-							MINIMUM_DEPOSIT,
-						)
-							.encode()
-							.into(),
-					},
-					ExpectTransactStatus(MaybeErrorCode::Success),
-				]);
-				instructions.into()
-			},
-			Command::SendToken { gateway, token, destination, amount } => {
-				let asset = MultiAsset::from((
-					Self::convert_token_address(network, gateway, token),
-					amount,
-				));
-
-				let origin_location = Junction::AccountKey20 { network: None, key: gateway.into() };
-
-				let mut instructions = create_instructions(origin_location);
-				instructions
-					.extend(vec![ReserveAssetDeposited(vec![asset.clone()].into()), ClearOrigin]);
-
-				let (dest_para_id, beneficiary) = match destination {
-					Destination::AccountId32 { id } => (
-						None,
-						MultiLocation {
-							parents: 0,
-							interior: X1(AccountId32 { network: None, id }),
-						},
-					),
-					Destination::ForeignAccountId32 { para_id, id } => (
-						Some(para_id),
-						MultiLocation {
-							parents: 0,
-							interior: X1(AccountId32 { network: None, id }),
-						},
-					),
-					Destination::ForeignAccountId20 { para_id, id } => (
-						Some(para_id),
-						MultiLocation {
-							parents: 0,
-							interior: X1(AccountKey20 { network: None, key: id }),
-						},
-					),
+pub type CallIndex = [u8; 2];
+
+impl<G: Get<CallIndex>> ConvertMessage for VersionedMessageToXcmConverter<G> {
+	fn convert(message: VersionedMessage) -> Result<Xcm<()>, ConvertMessageError> {
+		match message {
+			VersionedMessage::V1(val) => {
```

Comment:
The name `val` isn't great. I think let's destructure it:

```suggestion
			VersionedMessage::V1(MessageV1 { chain_id, message }) => {
```